modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/WeightedProspect.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.sdk.infrastructure.snapshot
+
+import cats.MonadThrow
+import cats.data.Chain
+import cats.effect.std.Random
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.trust.TrustValueRefined
+import org.tessellation.sdk.infrastructure.snapshot.WeightedSampling.{
+  ProbabilityRefinement,
+  Weight,
+  WeightRefinement,
+  sample => weightedSample
+}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.refineV
+import eu.timepit.refined.types.numeric.PosInt
+
+object WeightedProspect {
+
+  def sample[F[_]: MonadThrow: Random, A](candidates: Map[A, TrustValueRefined], sampleSize: PosInt): F[List[A]] = {
+    val filtered = candidates.view
+      .mapValues(score => refineV[WeightRefinement](score.value))
+      .collect {
+        case (key, Right(refinedScore)) =>
+          key -> refinedScore
+      }
+      .toMap
+
+    if (filtered.size <= sampleSize)
+      filtered.keySet.toList.pure
+    else {
+      type Result = List[A]
+      type Agg = (Chain[A], Map[A, Weight])
+
+      val getP = Random[F].nextDouble.map(refineV[ProbabilityRefinement].unsafeFrom(_))
+
+      (Chain.empty[A], filtered).tailRecM {
+        case (chosen, _) if chosen.size == sampleSize.value => chosen.toList.asRight[Agg].pure
+        case (chosen, d) =>
+          weightedSample(getP)(d.toList).map(c => (chosen :+ c, d - c).asLeft[Result])
```

Comment:
I hold my previous comment that it should be: `getP.flatMap(weightedSample(_)(d.toList))` so there is no need to pass `F[..]` as an argument to `weightedSample`

Reply:
Re-factored `WeightedSample.sample` to

```
def sample[F[_]: MonadThrow, A](chosenP: Probability)(distribution: List[(A, Weight)]): F[A]
```

and re-factored `WeightedProspect` and the tests accordingly.